helix-lock/src/main/java/org/apache/helix/lock/helix/ZKDistributedNonblockingLock.java

```diff
@@ -144,19 +305,143 @@ public ZNRecord update(ZNRecord current) {
+      throw new HelixException(
+          String.format("User %s failed to acquire lock at Lock path %s.", _userId, _lockPath));
     }
   }
 
+  /**
+   * Class that specifies how a lock node should be updated after the previous owner finishes the
+   * cleanup work
+   */
+  private class CleanupUpdater implements DataUpdater<ZNRecord> {
+    public CleanupUpdater() {
+    }
+
+    @Override
+    public ZNRecord update(ZNRecord current) {
+      // If we are still the lock owner, clean owner field.
+      LockInfo curLockInfo = new LockInfo(current);
+      if (isCurrentOwner(curLockInfo)) {
+        ZNRecord record = current;
+        record
+            .setSimpleField(LockInfo.LockInfoAttribute.OWNER.name(), LockConstants.DEFAULT_USER_ID);
+        record.setSimpleField(LockInfo.LockInfoAttribute.MESSAGE.name(),
+            LockConstants.DEFAULT_MESSAGE_TEXT);
+        record.setLongField(LockInfo.LockInfoAttribute.TIMEOUT.name(),
+            LockConstants.DEFAULT_TIMEOUT_LONG);
+        record.setIntField(LockInfo.LockInfoAttribute.PRIORITY.name(),
+            LockConstants.DEFAULT_PRIORITY_INT);
+        record.setLongField(LockInfo.LockInfoAttribute.WAITING_TIMEOUT.name(),
+            LockConstants.DEFAULT_WAITING_TIMEOUT_LONG);
+        record.setLongField(LockInfo.LockInfoAttribute.CLEANUP_TIMEOUT.name(),
+            LockConstants.DEFAULT_CLEANUP_TIMEOUT_LONG);
+        return record;
+      }
+      LOG.error("User {} is not current lock owner, and does not need to unlock {}", _userId,
+          _lockPath);
+      throw new HelixException(String
+          .format("User %s is not current lock owner, and does not need" + " to unlock %s", _userId,
+              _lockPath));
+    }
+  }
+
+  /**
+   * Class that specifies how a lock node should be updated during a forceful get lock operation
+   */
+  private class ForcefulUpdater implements DataUpdater<ZNRecord> {
+    final ZNRecord _record;
+
+    /**
+     * Initialize a structure for lock user to update a lock node value
+     * @param lockInfo the lock node value will be used to update the lock
+     */
+    public ForcefulUpdater(LockInfo lockInfo) {
+      _record = lockInfo.getRecord();
+    }
+
+    @Override
+    public ZNRecord update(ZNRecord current) {
+      // If we are still the lock requestor, change ourselves to be lock owner.
+      LockInfo curLockInfo = new LockInfo(current);
+      if (curLockInfo.getRequestorId().equals(_userId)) {
+        return _record;
+      }
+      LOG.error("User {} is not current lock requestor, and cannot forcefully acquire the lock at "
+          + "{}", _userId, _lockPath);
+      throw new HelixException(String.format("User %s is not current lock requestor, and cannot "
+          + "forcefully acquire the lock at %s", _userId, _lockPath));
+    }
+  }
+
+  private ZNRecord composeNewRequestorRecord(LockInfo oldLockInfo, ZNRecord newLockZNRecord) {
```

Comment:
nit, 
oldLockInfo => existingLockInfo, and newLockZNRecord => requestorLockZNRecord ?